### PR TITLE
AtomIdentifier (AID)

### PR DESCRIPTION
**Pull request by Sajjon**
PR for [this JiraTicket](https://radixdlt.atlassian.net/browse/RLAU-1320)

We introduce the `AtomIdentifier` which is created from an Atom's hash and `shards`.

### CODE REVIEW

File: RadixSDK.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -20,6 +20,8 @@
 		E6166BED2260CD5D003A9442 /* Port.swift in Sources */ = {isa = PBXBuildFile; fileRef = E6166BEC2260CD5D003A9442 /* Port.swift */; };
 		E620AB4821E785D600B36EC2 /* RadixSDK.h in Headers */ = {isa = PBXBuildFile; fileRef = E620AB4621E785D600B36EC2 /* RadixSDK.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		E620AB5921E7877A00B36EC2 /* PrivateKeyTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = E620AB5821E7874300B36EC2 /* PrivateKeyTests.swift */; };
+		E621BE832298394D00E97542 /* AtomIdentifier.swift in Sources */ = {isa = PBXBuildFile; fileRef = E621BE822298394D00E97542 /* AtomIdentifier.swift */; };
```
Comment (flotothemoon) [inline RadixSDK.xcodeproj/project.pbxproj]: Are these project files? Do they need to be in the repo?

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift [hunk 2]
```diff
@@ -8,9 +8,9 @@
 
 import Foundation
 
-// swiftlint:disable colon
+// swiftlint:disable colon opening_brace
 
-/// EffectiveUserIdentifier
+/// EffectiveUserIdentifier holding a 128 bit int
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift]: EUID stands for Extended Unique ID, not Effective User Identifier
The file name is "EffictiveIdentifier", so that would have to be updated too

File: Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift [hunk 3]
```diff
@@ -0,0 +1,207 @@
+//
+//  AtomIdentifierTests.swift
+//  RadixSDK iOS Tests
+//
+//  Created by Alexander Cyon on 2019-05-24.
```
Comment (flotothemoon) [inline Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift]: Over on the Java (Core + Libs) side of things we have decided not to include author names with files as the idea is that many people will contribute to any single file over time. Not sure if that applies to you, but I would personally like it to be consistent across projects

File: Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        let transferTokens = TransferTokenAction(from: alice, to: bob, amount: 10, tokenResourceIdentifier: rri)
+        
+        let consumables = createTokenAtom.spunParticles().compactMap { $0.mapToSpunParticle(with: TransferrableTokensParticle.self) }
+        XCTAssertEqual(consumables.count, 1)
+        let startBalance = TokenBalance(spunTransferrable: consumables[0])
+        
+        let atomFromTransfer = testAidOfAtomFrom(
+            action: transferTokens,
+            mapper: StatefulTransferTokensParticleGroupMapper(currentBalance: startBalance),
+            expectedAidShard: .eitherIn([alice.shard, bob.shard])
```
Comment (flotothemoon) [inline Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift]: It shouldn't be either alice or bobs shard but always the same

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+//
+//  Created by Alexander Cyon on 2019-05-24.
+//  Copyright © 2019 Radix DLT. All rights reserved.
+//
+
+import Foundation
+
+// swiftlint:disable colon opening_brace
+
+/// An Atom ID, made up of 192 bits of truncated hash and 64 bits of a selected shard. The Atom ID is used so that Atoms can be located using just their HashId (HID).
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: The Atom ID is used so that Atoms can be located using just the **Atom Identifier (AID)**, **not** HID

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    StringInitializable,
+    Comparable,
+    CustomStringConvertible
+{
+    // swiftlint:enable colon opening_brace
+    
+    /// 192 first bits of a `RadixHash`
+    private let truncatedHash: Data
+    
+    /// 64 bits of a
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: of a shard?

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+public extension AtomIdentifier {
+    static func == (lhs: AtomIdentifier, rhs: AtomIdentifier) -> Bool {
+        return lhs.asData == rhs.asData
+    }
+}
+
+// MARK: - Comparable
+public extension AtomIdentifier {
+    static func < (lhs: AtomIdentifier, rhs: AtomIdentifier) -> Bool {
+        return lhs.hex < rhs.hex
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Is it necessary to make this comparable? 
AID does not seem like something you can intuitively order to me.. how does this work with comparators and comparables in Swift? You don't need to implement Comparable to sort something, you can just use a Comparator at the site of sorting, right?

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    /// Deterministically selects a shard from a set of potential shards for any given Atom, using the RadixHash of said Atom.
+    ///
+    /// This function is defined in the Tempo whitepaper under the section about `AID` as:
+    /// ```
+    /// s(hash, shardSet) = shardSet.elementAt(hash.firstByte % shardSet.size)
+    /// ```
+    ///
+    static func selectShard(in shards: Shards, basedOnHash hash: RadixHash) -> Shard {
+        let targetShardIndex = Int(hash[0]) % shards.count
+        return shards.sorted()[targetShardIndex]
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: As discussed, need to pass in an **unsigned** comparator here.. my fault for not making this clearer

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift [hunk 9]
```diff
@@ -19,61 +19,79 @@ public struct EUID:
...
 public extension EUID {
-    var dsonPrefix: DSONPrefix {
-        return .euidHex
+
+    init(value: Value) throws {
+        try self.init(data: value.toData(minByteCount: EUID.byteCount))
+    }
+    
+    /// If a positive integer value is passed that will be used directly, if a negative int is passed, the the two-s complement of that will be used.
+    init(int: Int) throws {
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift]: Why twos complement on negative? Why even allow negative values at all?

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+public extension AtomIdentifier {
+    
+    static let byteCountHash = 24
+    static let byteCountShard = 8
+    static let byteCount = byteCountHash + byteCountShard
+}
+
+// MARK: - Throwing
+public extension AtomIdentifier {
+    enum Error: Swift.Error {
```
Comment (flotothemoon) [inline Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Uh, this is nice!
Kind of reminds me of Rusts error handling system

File: RadixSDK.xcodeproj/project.pbxproj (hunk 1 above)
Comment (Sajjon) [reply RadixSDK.xcodeproj/project.pbxproj]: Yes required by Xcode unfortunately :/

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift (hunk 2 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift]: Thanks for this! :D

File: Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift (hunk 3 above)
Comment (Sajjon) [reply Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift]: Good remark, I can change that in a separate PR.

File: Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift (hunk 4 above)
Comment (Sajjon) [reply Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift]: @flotothemoon No that is not true, since the `TransferTokenAction` results in an atom containing `TransferrableTokensParticle`, which contains a `Nonce`, resulting in different RadixHash of said Atom, yielding different `shard`, here is the JSON of the Atom created from the `TransferTokenAction`:

```json
{
  "metaData": {
    "timestamp": ":str:236"
  },
  "particleGroups": [
    {
      "particles": [
        {
          "particle": {
            "address": ":adr:JFeqmatdMyjxNce38w3pEfDeJ9CV6NCkygDt3kXtivHLsP3p846",
            "amount": ":u20:10",
            "granularity": ":u20:1",
            "nonce": -5792480063681574173,
            "permissions": {
              "burn": ":str:none",
              "mint": ":str:token_creation_only"
            },
            "planck": 26044039,
            "serializer": "radix.particles.transferrable_tokens",
            "tokenDefinitionReference": ":rri:/JF5FTU5wdsKNp4qcuFJ1aD9enPQMocJLCqvHE2ZPDjUNag8MKun/TEST",
            "version": 100
          },
          "serializer": "radix.spun_particle",
          "spin": 1,
          "version": 100
        },
        {
          "particle": {
            "address": ":adr:JF5FTU5wdsKNp4qcuFJ1aD9enPQMocJLCqvHE2ZPDjUNag8MKun",
            "amount": ":u20:10",
            "granularity": ":u20:1",
            "nonce": 7523508372601140568,
            "permissions": {
              "burn": ":str:none",
              "mint": ":str:token_creation_only"
            },
            "planck": 26044039,
            "serializer": "radix.particles.transferrable_tokens",
            "tokenDefinitionReference": ":rri:/JF5FTU5wdsKNp4qcuFJ1aD9enPQMocJLCqvHE2ZPDjUNag8MKun/TEST",
            "version": 100
          },
          "serializer": "radix.spun_particle",
          "spin": -1,
          "version": 100
        }
      ],
      "serializer": "radix.particle_group",
      "version": 100
    }
  ],
  "serializer": "radix.atom",
  "version": 100
}
```

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 5 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: @flotothemoon I copied this comment from the doc of `class AID` from the Java library, which means that that comment is wrong then, will you change that?

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 6 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: :)

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 8 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Great catch! If the Java lib unit tests would have contained shards with negative values this would have been caught by me :) So I will add uni tests with negative shards and cross compare with Java lib.

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift (hunk 9 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/EffictiveIdentifier (EUID)/EUID.swift]: Java lib did it, so followed that as a reference :)

File: Source/Stack/5. Layer 2 - ATOM MODEL (Particles)/Address/Shards/Shards.swift [hunk 11]
```diff
@@ -2,76 +2,10 @@
...
-    }
-    
-    func encodableKeyValues() throws -> [EncodableKeyValue<Shards.CodingKeys>] {
-        return [
-            EncodableKeyValue(key: .low, value: range.lowerBound),
-            EncodableKeyValue(key: .high, value: range.upperBound)
-        ]
-    }
-}
+public typealias Shards = NonEmptySet<Shard>
```
[the remark this answers is not in the document]
Comment (Sajjon) [reply Source/Stack/5. Layer 2 - ATOM MODEL (Particles)/Address/Shards/Shards.swift]: @mpsandiford I use rule 3 and 4 of Object Callestinics through out the lib:
https://williamdurand.fr/2013/06/03/object-calisthenics/#4-first-class-collections

Whenever I have an instance of `NonEmptySet`, then I know it is not empty, allowing me for skiiping these kind of `if` statement, from java lib:

```java
public static AID from(RadixHash hash, Set<Long> shards) {
    ....
    if (shards.isEmpty()) {
	throw new IllegalArgumentException("Shards cannot be empty");
    }
```

File: Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift (hunk 4 above)
Comment (flotothemoon) [reply Tests/TestCases/Radix/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/SubatomicModels/AtomIdentifier/AtomIdentifierTests.swift]: Woah, you're totally right! I somehow assumed it would be the same atom every time since I tend to assume static inputs in tests.

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 7 above)
Comment (flotothemoon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: To clue Alex in, Martin's laughing because he brought this up in my PR in Core.. :D

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 5 above)
Comment (flotothemoon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Ah shit. Yes that needs updating

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 7 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Haha, I should not so blindly follow the Java code ;) I will remove `Comparable`

File: Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift (hunk 10 above)
Comment (Sajjon) [reply Source/Stack/6. Layer 1 - SUBATOMIC (Serialization, Crypto, Models)/Models/Subatomic Particles/AtomIdentifier (AID)/AtomIdentifier.swift]: Thanks to your comment I realized that the case `shardSetCannotBeEmpty` (Solved by the dedicated type `Shards`, which cannot be instantiated using an empty set. Thus error thrown in the init of `Shards`.) is no longer needed, thanks!

File: Source/Stack/5. Layer 2 - ATOM MODEL (Particles)/Address/Shards/Sharded.swift [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+//  Sharded.swift
+//  RadixSDK iOS
+//
+//  Created by Alexander Cyon on 2019-05-27.
+//  Copyright © 2019 Radix DLT. All rights reserved.
+//
+
+import Foundation
+
+public typealias Shard = Int64
```
[the remark this answers is not in the document]
Comment (Sajjon) [reply Source/Stack/5. Layer 2 - ATOM MODEL (Particles)/Address/Shards/Sharded.swift]: @mpsandiford The reason why int64 is used is because it is used by `ShardInterval` type, which is part of the `RadixSystem` type, returned by `NodeInfo` call ([API docs here](https://docs.radixdlt.com/node-api/rest#get-get-live-peers)):

```json
  {
...
      "shards": {
        "high": 9223372036854775807,
        "low": -9223372036854775808
      },
...
```

And these shards are signed as you can see.

So sometimes `Shard` is Int64, and sometimes UInt64? Is this not confusing, can you shed some light here? :)

**Comment by Sajjon** [comment]
@flotothemoon Fixed everything now, I think :)

Based my Swift unit tests on some Java unit tests I wrote for shard sets with negative values:

```java
private void doTestShards(
			int firstByteInHash,
			Set<Long> shards,
			List<Long> expectedOrderWhenSorted,
			long expectedShard
) {
	byte[] hashBytes = new byte[Hash.BYTES];
	for (int i = 0; i < Hash.BYTES; i++) {
		hashBytes[i] = (byte) (i + firstByteInHash);
	}
	RadixHash hash = mock(RadixHash.class);
	when(hash.getFirstByte()).thenReturn(hashBytes[0]);
	doAnswer((Answer<Object>) invocation -> {
		byte[] destination = (byte[]) invocation.getArguments()[0];
		System.arraycopy(hashBytes, 0, destination, 0, AID.HASH_BYTES);
		return null;
	}).when(hash).copyTo(new byte[0], 0, 0);

	List<Long> sortedShards = shards.stream().sorted(Long::compareUnsigned).collect(Collectors.toList());

	String expectedOrderingDescription = String.format("When `%s` is sorted (skipping sign), we expect it to equal `%s`, but got: `%s`", shards, expectedOrderWhenSorted, sortedShards);

	assertEquals(expectedOrderingDescription, expectedOrderWhenSorted, sortedShards);

	String expectedShardDescription = String.format("firstByteInHash of `%d`, and shardSet size of `%d`, yields target index: `%s`", firstByteInHash, shards.size(), firstByteInHash % shards.size());

	assertEquals(expectedShardDescription, expectedShard, AID.from(hash, shards).getShard());
}


@Test
public void testShardSets() {
	doTestShards(
			5,
			ImmutableSet.of(9L, -7L, 5L),
			ImmutableList.of(5L, 9L, -7L),
			-7L // 5 % 3 => index 2 => -7
	);

	doTestShards(
			6,
			ImmutableSet.of(9L, -7L, 5L),
			ImmutableList.of(5L, 9L, -7L),
			5L // 6 % 3 => index 0 => 5L
	);


	doTestShards(
			7,
			ImmutableSet.of(9L, -7L, 5L),
			ImmutableList.of(5L, 9L, -7L),
			9L // 7 % 3 => index 1 => 9L
	);

	doTestShards(
			9,
			ImmutableSet.of(-9L, -7L, -13L, 5L),
			ImmutableList.of(5L, -13L, -9L, -7L),
			-13L // 9 % 4 => index 1 => -13
	);
}
```
